### DISCUSSION TITLE
Is there a symbol for aot?

### DISCUSSION THREAD

**Discussion by WeihanLi** [dotnet/runtime#109005] [Q&A] [answered]
I'd like to write some code excluded for AOT, wondering if there's something like `#if AOT` to write conditional code for AOT only or none AOT only?

**Comment by huoyaoyuan** [comment, marked as answer]
You can use `RuntimeFeature.IsDynamicCodeSupported` as a feature guard. Preprocessors won't work here because AOT uses the same C# compilation.

**Reply by WeihanLi** [reply]
get it, thanks very much

**Reply by am11** [reply]
`IsDynamicCodeSupported` is not reliable either: https://github.com/dotnet/BenchmarkDotNet/pull/2561#issuecomment-2054119721. If you want to check if PublishAot=true, you can propagate it via `<DefineConstants Condition="'$(PublishAot)' == 'true'">$(DefineConstants);NATIVEAOT</DefineConstants>` and check `#if NATIVEAOT`, like runtime repo is doing:

https://github.com/dotnet/runtime/blob/be013ed306ea178907baca6763601f9b2e69a400/src/coreclr/nativeaot/Directory.Build.props#L43
https://github.com/dotnet/runtime/blob/be013ed306ea178907baca6763601f9b2e69a400/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadBlockingInfo.cs#L115

**Reply by vitek-karas** [reply]
In general we recommend to try
* Write the code in an AOT compatible way for every target (the core libraries should take of most of this for you)
* Determine what exact functionality you rely on (dynamic code generation, assembly files, ...) and use the appropriate feature check like IsDynamicCodeSupported.
* If it's not possible to make the code AOT compatible at all, decorate it with `RequiresDynamicCodeAttribute` so that whoever calls it is notified that it will not work with AOT.
